daemon/containerd/image.go

```diff
@@ -2,14 +2,222 @@ package containerd
-	return nil, errdefs.NotImplemented(errors.New("not implemented"))
+func (i *ImageService) GetImage(ctx context.Context, refOrID string, options imagetype.GetImageOpts) (*image.Image, error) {
+	desc, err := i.resolveDescriptor(ctx, refOrID)
+	if err != nil {
+		return nil, err
+	}
+
+	platform := platforms.AllPlatformsWithPreference(cplatforms.Default())
+	if options.Platform != nil {
+		platform = cplatforms.OnlyStrict(*options.Platform)
+	}
+
+	cs := i.client.ContentStore()
+	conf, err := containerdimages.Config(ctx, cs, desc, platform)
+	if err != nil {
+		return nil, err
+	}
+
+	imageConfigBytes, err := content.ReadBlob(ctx, cs, conf)
+	if err != nil {
+		return nil, err
+	}
+
+	var ociimage ocispec.Image
+	if err := json.Unmarshal(imageConfigBytes, &ociimage); err != nil {
+		return nil, err
+	}
+
+	rootfs := image.NewRootFS()
+	for _, id := range ociimage.RootFS.DiffIDs {
+		rootfs.Append(layer.DiffID(id))
+	}
+	exposedPorts := make(nat.PortSet, len(ociimage.Config.ExposedPorts))
+	for k, v := range ociimage.Config.ExposedPorts {
+		exposedPorts[nat.Port(k)] = v
+	}
+
+	img := image.NewImage(image.ID(desc.Digest))
+	img.V1Image = image.V1Image{
+		ID:           string(desc.Digest),
+		OS:           ociimage.OS,
+		Architecture: ociimage.Architecture,
+		Config: &containertypes.Config{
+			Entrypoint:   ociimage.Config.Entrypoint,
+			Env:          ociimage.Config.Env,
+			Cmd:          ociimage.Config.Cmd,
+			User:         ociimage.Config.User,
+			WorkingDir:   ociimage.Config.WorkingDir,
+			ExposedPorts: exposedPorts,
+			Volumes:      ociimage.Config.Volumes,
+			Labels:       ociimage.Config.Labels,
+			StopSignal:   ociimage.Config.StopSignal,
+		},
+	}
+
+	img.RootFS = rootfs
+
+	if options.Details {
+		lastUpdated := time.Unix(0, 0)
+		size, err := i.size(ctx, desc, platform)
+		if err != nil {
+			return nil, err
+		}
+
+		tagged, err := i.client.ImageService().List(ctx, "target.digest=="+desc.Digest.String())
+		if err != nil {
+			return nil, err
+		}
+		tags := make([]reference.Named, 0, len(tagged))
+		for _, i := range tagged {
+			if i.UpdatedAt.After(lastUpdated) {
+				lastUpdated = i.UpdatedAt
+			}
+			name, err := reference.ParseNamed(i.Name)
+			if err != nil {
+				return nil, err
+			}
+			tags = append(tags, name)
+		}
+
+		img.Details = &image.Details{
+			References:  tags,
+			Size:        size,
+			Metadata:    nil,
+			Driver:      i.snapshotter,
+			LastUpdated: lastUpdated,
+		}
+	}
+
+	return img, nil
+}
+
+// size returns the total size of the image's packed resources.
+func (i *ImageService) size(ctx context.Context, desc ocispec.Descriptor, platform cplatforms.MatchComparer) (int64, error) {
+	var (
+		cs      = i.client.ContentStore()
+		handler = containerdimages.ChildrenHandler(cs)
+		size    int64
+	)
+
+	handler = containerdimages.LimitManifests(handler, platform, 1)
```

Comment:
(No need to change if there's no other issues remaining); I noticed this is wrapping the `handler` that's defined just a few lines above; could probably define it all at once here;

```go
var (
	cs      = i.client.ContentStore()
	handler = containerdimages.LimitManifests(containerdimages.ChildrenHandler(cs), platform, 1)
	size    int64
)
```


or

```go
var size int64

cs := i.client.ContentStore()
handler := containerdimages.LimitManifests(containerdimages.ChildrenHandler(cs), platform, 1)
```